rasa/core/events/__init__.py

```diff
@@ -632,87 +651,143 @@ def __eq__(self, other) -> bool:
+                parameters.get("entities"),
+                name=parameters.get("name"),
+                kill_on_user_message=parameters.get("kill_on_user_msg", True),
+                timestamp=parameters.get("timestamp"),
+                metadata=parameters.get("metadata"),
             )
         ]
 
 
 # noinspection PyProtectedMember
 class ReminderCancelled(Event):
-    """Cancel all jobs with a specific name."""
+    """Cancel certain jobs."""
 
     type_name = "cancel_reminder"
 
     def __init__(
         self,
-        action_name: Text,
+        name: Optional[Text] = None,
+        intent: Optional[Text] = None,
+        entities: Optional[List[Dict]] = None,
         timestamp: Optional[float] = None,
         metadata: Optional[Dict[Text, Any]] = None,
     ):
-        """
+        """Creates a ReminderCancelled event.
+
+        All arguments are optional and used to refine the set of reminders that
+        are to be cancelled. If no arguments are supplied, this will cancel all reminders.
+
         Args:
-            action_name: name of the scheduled action to be cancelled
-            metadata: optional event metadata
+            name: Name of the reminder to be cancelled.
+            intent: Intent name that is to be used to identify the reminders to be cancelled.
+            entities: Entities that are to be used to identify the reminders to be cancelled.
+            timestamp: Optional timestamp.
+            metadata: Optional event metadata.
         """
 
-        self.action_name = action_name
+        self.name = name
+        self.intent = intent
+        self.entities = entities
         super().__init__(timestamp, metadata)
 
     def __hash__(self) -> int:
-        return hash(self.action_name)
+        return hash((self.name, self.intent, str(self.entities),))
 
     def __eq__(self, other) -> bool:
-        return isinstance(other, ReminderCancelled)
+        if not isinstance(other, ReminderCancelled):
+            return False
+        else:
+            return hash(self) == hash(other)
 
     def __str__(self) -> Text:
-        return f"ReminderCancelled(action: {self.action_name})"
+        return f"ReminderCancelled(name: {self.name}, intent: {self.intent}, entities: {self.entities})"
+
+    def cancels_job_with_name(self, job_name: Text, sender_id: Text) -> bool:
+        """Determines if this ReminderCancelled should cancel the job with the given name.
+
+        Args:
+            job_name: Name of the job to be tested.
+            sender_id: The `sender_id` of the tracker.
+
+        Returns:
+            `True`, if this `ReminderCancelled` should cancel the job with the given name,
+            and `False` otherwise.
+        """
+
+        match = re.match(
+            rf"^\[([\d\-]*),([\d\-]*),([\d\-]*)\]"
+            + rf"({re.escape(ACTION_NAME_SENDER_ID_CONNECTOR_STR)}{re.escape(sender_id)})",
+            job_name,
+        )
+        if not match:
```

Comment:
we should still check if we actually matched these groups so that we can fail gracefully. 
It's 4 groups in total, right? (best write a test)
```suggestion
        if not match or match.lastindex < 3:
```

Reply:
If the groups are not matched, `match` will be `none`. It should not be possible for `match` to be not `none` and not have `match.lastindex == 4`. Either the pattern is satisfied, or it is not.

Reply:
I mean the only way you'd get `match.lastindex != 4` is if you change the regex, since there is currently no optional pattern in there. I think adding `or match.lastindex < 4` would just be confusing, since it suggests that it could be different.

Reply:
True, thanks for clarifying!